rate_limiter.rst

```diff
@@ -0,0 +1,187 @@
+resets and you have another 100 requests for the following 60 minutes.
+
+In the ``authenticated_api`` limiter, after making the first HTTP request you
+are allowed to make up to 5,000 HTTP requests in total, and this number grows
+at a rate of another 5,000 requests per hour. If you don't make that number of
+requests, the unused ones don't accumulate (the ``limit`` option prevents that
+number from being higher than 5,000).
+
+Rate Limiting in Action
+-----------------------
+
+After having installed and configured the rate limiter, inject it in any service
+or controller and call the ``consume()`` method to try to consume a given number
+of tokens. For example, this controller uses the previous rate limiter to control
+the number of requests to the API::
+
+    // src/Controller/ApiController.php
+    namespace App\Controller;
+
+    use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+    use Symfony\Component\HttpKernel\Exception\TooManyRequestsHttpException;
+    use Symfony\Component\RateLimiter\LimiterInterface;
+
+    class ApiController extends AbstractController
+    {
+        // the variable name must be: "rate limiter name" + "limiter" suffix
+        public function index(LimiterInterface $anonymousApiLimiter)
+        {
+            // create a limiter based on a unique identifier of the client
+            // (e.g. the client's IP address, a username/email, an API key, etc.)
+            $limiter = $anonymousApiLimiter->create($request->getClientIp());
+
+            // the argument of consume() is the number of tokens to consume
+            // and returns an object of type Limit
+            if (false === $anonymous_api_limiter->consume(1)->isAccepted()) {
+                throw new TooManyRequestsHttpException();
+            }
+
+            // you can also use the ensureAccepted() method - which throws a
+            // RateLimitExceededException if the limit has been reached
+            // $limiter->consume(1)->ensureAccepted();
+
+            // ...
+        }
+
+        // ...
+    }
+
+.. note::
+
+    In a real application, instead of checking the rate limiter in all the API
+    controller methods, create an :doc:`event listener or subscriber </event_dispatcher>`
+    for the :ref:`kernel.request event <component-http-kernel-kernel-request>`
+    and check the rate limiter once for all requests.
+
+In other scenarios you may want instead to wait as long as needed until a new
+token is available. In those cases, use the ``reserve()`` and ``wait()`` methods::
+
+    // src/Controller/LuckyController.php
+    // ...
+
+    class ApiController extends AbstractController
+    {
+        public function registerUser(LimiterInterface $authenticatedApiLimiter)
+        {
+            $limiter = $authenticatedApiLimiter->create($request->getClientIp());
+
+            // this blocks the application until the given number of tokens can be consumed
+            $limiter->consume(1)->wait();
```

Comment:
```suggestion
            // this blocks the application until the given number of tokens can be consumed
            do {
                $limit = $limiter->consume(1);
                $limit->wait();
            } while (!$limit->isAccepted());
```

The downside of `consume()` is that there is no guarantee that you are accepted after `wait()`. E.g. if the window allows 5 requests per minute and 6 processes are blocked, 1 process is not accepted after the window reset and has to wait at least another minute.

Idea: add a helper method for this with a `$maxWaitTime`.